### ISSUE TITLE
PGCoord fails to get pubsub updates

### ISSUE THREAD

**Issue by spikecurtis** [coder/coder#11950]
reproduced on our dogfood instance.

One of our 3 replicas is no longer receiving peer and tunnel updates, so if your agent or client connects to that coordinator, it will fail to set up networking.  I can see this via debug logging: 2 of 3 coordinators show peer and tunnel updates but the 3rd coordinator drops no logs about getting updates.

When I connected from a client and landed on the bad coordinator, my client never got the agent node update, but the agent, connected to a "good" coordinator did.

**Comment by spikecurtis** [comment]
The broken instance gets some updates for like 10ms right at start of day, then doesn't get any more updates.  Logs in and around that time seem fine.

SQL logs and connections look fine.

Unfortunately our pbPubsub is entirely uninstrumented: no logs, no metrics.

**Comment by spikecurtis** [comment]
After redeploy, everything seems to be working fine.  I'm adding instrumentation to pubsub, and hopefully we'll see a recurrence that the new instrumentation will help us understand.

**Comment by spikecurtis** [comment]
Given we have seen this from multiple customers, but have yet to reproduce it in our own infrastructure since the logging & metrics upgrades to pubsub, I wonder how people would feel about a pubsub "watchdog" that periodically sends a heartbeat on a well known channel, subscribes to that channel, and then kills the whole process if we don't get messages for 3x the heartbeat period. (I'm thinking 15s period, thus 45s of zombie pubsub before we self-immolate)

It's definitely a workaround, rather than a fix. And, killing the process is fairly extreme. Anecdotally, we believe that restarting the process restores functionality, and a coderd without a working pubsub causes all kinds of problems.  Before killing the process I'd want to dump goroutine stack traces to the log (_a al_ the tailscale engine watchdog) so that if the pubsub is due to some deadlock in the process we'll have a prayer of finding it.

@kylecarbs @bpmct @ammario @sreya @mtojek I'm interested in your thoughts here.

**Comment by bpmct** [comment]
We're doing to set up alerts on our test instances to see if this occurs

**Comment by spikecurtis** [comment]
The watchdog triggered and we got a stack trace.

tl;dr `PGPubsub` can enter into a deadlock with `pq.Listener` based on the way channels and locking works.

------

PGPubsub is protected by a single global mutex that is held (among other times) while 

a) subscribing to a new event stream and
b) delivering notifications

The `pq.Listener` uses a single mainloop to read postgres responses both about subscriptions and notifications.  The deadlock works like this:

1. A new request to subscribe comes into PGPubsub and we lock the mutex
2. While holding the mutex, we call `pq.(*Listener).Listen()` which sends a request to postgres and then waits for a response.
3. Meanwhile, the `PGPubsub` attempts to deliver a notification, and blocks attempting to acquire the mutex.
4. Meanwhile, the `pq.Listener()` main loop receives a notification, and attempts to send it over a channel to PGPubsub, but this blocks because the goroutine in 3 is blocked.  The goroutines in 2, 3, 4 are deadlocked and cannot make forward progress.

-----

A relatively simple fix is to release the mutex while we call `pq.(*Listener).Listen()`.  This should allow the other goroutines to make forward progress and avoid the deadlock.
